File: Cpp/46.cpp
//经典的全排列
//回溯法
class Solution {
public:
    vector<vector<int>> res;
    vector<int> path;

    void backtrack(vector<int> &nums, vector<bool> &used) {
        if(path.size() == nums.size()) {
            res.push_back(path);
            return;
        }
        for(int i=0; i<nums.size(); ++i) {
            if(used[i]) continue;
            used[i] = true;
            path.push_back(nums[i]);
            backtrack(nums,used);
            path.pop_back();
            used[i] = false;
        }
    }

    vector<vector<int>> permute(vector<int>& nums) {
        vector<bool> used(nums.size(), false);
        backtrack(nums, used);
        return res;
    }
};

//参考：
//https://leetcode-cn.com/problems/permutations/solution/46-quan-pai-lie-hui-su-suan-fa-jing-dian-ti-mu-xia/

File: JavaScript/46.js
var permute = function(nums) {

    const backtrack = function(nums, used) {
        if(path.length === nums.length) {
            res.push(path.slice());
            // slice 方法返回一个新的数组对象
            // 这一对象是一个原数组的浅拷贝, 范围[begin, end)
            // 若begin省略默认从0开始, 若end省略默认拷贝到数组末尾
            return;
        }
        for(let i=0; i<nums.length; ++i) {
            if(used[i]) continue;
            path.push(nums[i]);
            used[i] = true;
            backtrack(nums, used);
            path.pop();
            used[i] = false;
        }
    }
    let res = [], path = [];
    backtrack(nums, []);
    return res;
};